runners/java-fn-execution/src/test/java/org/apache/beam/runners/fnexecution/logging/GrpcLoggingServiceTest.java

```diff
@@ -0,0 +1,207 @@
+                    createLogsWithIds(instructionReference, -instructionReference));
+                outboundObserver.onError(new RuntimeException("Client " + instructionReference));
+                waitForTermination.await();
+                return null;
+              }
+            });
+      }
+      ExecutorService executorService = Executors.newCachedThreadPool();
+      executorService.invokeAll(tasks);
+    }
+  }
+
+  @Test
+  public void testServerCloseHangsUpClients() throws Exception {
+    LinkedBlockingQueue<LogEntry> logs = new LinkedBlockingQueue<>();
+    ExecutorService executorService = Executors.newCachedThreadPool();
+    Collection<Future<Void>> futures = new ArrayList<>();
+    final GrpcLoggingService service =
+        GrpcLoggingService.forWriter(new ConcurrentCollectionLogWriter(logs));
+    try (GrpcFnServer<GrpcLoggingService> server =
+        GrpcFnServer.allocatePortAndCreateFor(service, InProcessServerFactory.create())) {
+
+      for (int i = 1; i <= 3; ++i) {
+        final long instructionReference = i;
+        futures.add(
+            executorService.submit(
+                new Callable<Void>() {
+                  public Void call() throws Exception {
+                    {
+                      CountDownLatch waitForServerHangup = new CountDownLatch(1);
+                      ManagedChannel channel =
+                          InProcessChannelBuilder.forName(
+                                  server.getApiServiceDescriptor().getUrl())
+                              .build();
+                      StreamObserver<BeamFnApi.LogEntry.List> outboundObserver =
+                          BeamFnLoggingGrpc.newStub(channel)
+                              .logging(
+                                  TestStreams.withOnNext(messageDiscarder)
+                                      .withOnCompleted(new CountDown(waitForServerHangup))
+                                      .build());
+                      outboundObserver.onNext(createLogsWithIds(instructionReference));
+                      waitForServerHangup.await();
+                      return null;
+                    }
+                  }
+                }));
+      }
+      // Wait till each client has sent their message showing that they have connected.
+      for (int i = 1; i <= 3; ++i) {
+        logs.take();
+      }
+    }
+    System.out.println("Server closed");
+    for (Future<Void> future : futures) {
+      future.get();
+    }
+  }
+
+  private BeamFnApi.LogEntry.List createLogsWithIds(long... ids) {
+    BeamFnApi.LogEntry.List.Builder builder = BeamFnApi.LogEntry.List.newBuilder();
+    for (long id : ids) {
+      builder.addLogEntries(createLogWithId(id));
+    }
+    return builder.build();
+  }
+  private BeamFnApi.LogEntry createLogWithId(long id) {
+    return BeamFnApi.LogEntry.newBuilder().setInstructionReference(Long.toString(id)).build();
+  }
+
+  private static class ConcurrentCollectionLogWriter implements LogWriter {
```

Comment:
Nothing enforces that this is a ConcurrentCollectionLogWriter, rename to CollectionAppendingLogWriter?

Reply:
Done.